ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/JumpingGovernor.hs

```diff
@@ -0,0 +1,337 @@
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE DeriveGeneric        #-}
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE NamedFieldPuns       #-}
+{-# LANGUAGE RankNTypes           #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE StandaloneDeriving   #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Ouroboros.Consensus.MiniProtocol.ChainSync.Client.JumpingGovernor (
+    ClientState
+  , Handle
+  , Instruction (..)
+  , JumpResult (..)
+  , dummyHandle
+  , newHandle
+  , nextInstruction
+  , processJumpResult
+  , registerClient
+  , unregisterClient
+  ) where
+
+import           Cardano.Slotting.Slot (SlotNo, WithOrigin (..))
+import           Control.Monad (forM_, when)
+import qualified Data.List as List
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromJust, mapMaybe)
+import           GHC.Generics (Generic)
+import           Ouroboros.Consensus.Block (HasHeader, Header, Point (..),
+                     WithOrigin (..), castPoint, succWithOrigin)
+import           Ouroboros.Consensus.MiniProtocol.ChainSync.Client.State
+                     (ChainSyncClientHandle (..), ChainSyncState (..))
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment,
+                     blockPoint, headPoint, headSlot, splitAfterPoint,
+                     splitBeforePoint, toOldestFirst)
+
+data Handle m peer blk = Handle
+  { clientHandles :: !(StrictTVar m (Map peer (ChainSyncClientHandle m blk))),
+    clientStates  :: !(StrictTVar m (Map peer (ClientState m blk)))
```

Comment:
This is an artifact of me trying to get things done before cleaning up. I have since then pushed the CSJ-specific state into `ChainSyncHandle`; cf 4be366d53504d63bd174d2c86d9db3f639b42c8f.